### ISSUE TITLE
[BUG] Text cut-off mid-generation with Chain of Thought prompts in SillyTavern

### ISSUE THREAD

**Issue by Tomolasthill** [SillyTavern/SillyTavern#2536]
### Environment

🪟 Windows

### System

Windows 11 V 23H2 (OS Build 22531.3380)

### Version

SillyTavern 1.12.3 'release' (3c1d639c)
SillyTavern 1.12.3 'staging' (21cdfa1d)

### Desktop Information

Node version: v20.13.0
Claude Chat Completion - Messages API (API Key via Anthropic)
Model: Claude 3 Family (Opus, Sonnet and Sonnet 3.5)

#### Update Note: *For sake of transparency and easier overview, it has been determined that his is not a model related issue and instead a SillyTavern function causing this, it was just easier to trigger this on Claude models then others. Discussion has been shifted to focus on the actual function, name and lack of information regarding what it does. This note is made instead of doing drastic changes to the opening post, as after discovering the actual reason for this issue, OP post feels a bit off-point.*

I believe this is a SillyTavern issue and not a model issue, as the problem in question works without issues on other models (GPT-4, Mistral, Gemini, etc.), but for Claude 3 models, it's failing too frequently.

The issue at hand, as per the title, is that when trying to generate a response with Claude 3 (Sonnet 3.5 or Opus), the response is cut off during generation, BUT the generation isn't finished. It continues to generate for a few seconds until it completes the response. However, it doesn't show the text. I have a video file to showcase this.

One thing to note is that this issue seems to be related to using Chain of Thought (CoT) prompts, specifically those that make use of 'special' formatting such as Markdown, XML tags, bullet points, etc. I've confirmed this by doing multiple generations without CoT, and this issue never occurred. However, the moment I enable a CoT, it begins to happen infrequently, with a near 100% failure rate on more complex CoTs.

Here is a [Streamable link](https://streamable.com/il99af) to the video showcasing a swipe that demonstrates the issue.

To further illustrate the issue, I'll be posting a CoT I've been using for roleplaying purposes, meant to have the Claude models stick to the character more. The CoTs are being injected as role "User" @ depth 0, so after the User's message.

<details>
 <summary>
 Large CoT
 </summary>

```
Begin with making an analysis of character(s) in scene. Assess accuracy, suggest fixes, rate 1-10. Use this template but GREATLY simplify it:

<Analysis>
### Here is the characters analysis.
Character X:
- Gender, personality, outfit, key traits

Rating: Y/10 - Rationale

Current action/motivation:
- ABC

How to improve authenticity:
- ABC

Escalation options (if narrative calls for it!):
- ABC

Quirks/mannerisms to deploy:
- Speech style (slurred/clear/muffled)
- Body language
- Focus areas
- Other notable behaviors

### Do the same of character Y.

Writing style:
- Type of language? (Articulate? Educated? Slurry? Accent? etc..)
- Which Author to mimic? (At least two)
</Analysis>

Mince no words and be concise and condensed.
```
</details>

Now, if I shorten the CoT to something like this:

<details>
 <summary>
 Shortened CoT
 </summary>

```
Analyze characters. Rate 1-10. Use this template:

<Analysis>
Character X:
- Basic traits
- Current action/motivation
- Improvement ideas
- Escalation options
- Desires
- Quirks

Character Y: [Repeat above]

Writing style:
- Language type
- Authors to mimic
</Analysis>

Be brief and direct.
```
</details>

It still cuts off the text, **however**, if I replace the `Character X` and `Character Y` with either `Character {{char}}` and `Character {{user}}` macro (or their actual respective names), then the CoT suddenly works! Meaning the text is no longer cut off mid generation for some reason.

Also, these CoTs are being run with the following prefill: `I, Claude, understand. I will now provide the analysis and follow the template.` Sometimes, regardless of the CoT, if I add `{{char}}:` to the prefill, it will often cause it to crash as well.

Unfortunately, I haven't been able to determine exactly why this happens. It's occurring too infrequently for me to pinpoint the cause. However, I do feel that the more 'complex' the CoT is (in terms of symbols, formatting, and markdown), the higher the likelihood of the text being cut off mid-generation.

This is quite annoying when the generations are so costly as the text is still being generated, SillyTavern is just not showing/recieving it. And yes, I have checked if the text is just hidden for some reason. Clicking Edit still shows nothing.

For reference. I'm using a fresh SillyTavern installation to test this. So it's not any additional features, extension or user modification that's causing this.

If you have any further question, I will try to reply as soon as I can.

### Additional info

Console log. Taken after a hard refresh and one generation which a CoT which cut off mid-generation.
<details>
 <summary>
 Log
 </summary>

```
expressions: SillyTavern is now hidden index.js:1983:21
Layout was forced before the page was fully loaded. If stylesheets are not yet loaded this may cause a flash of unstyled content. index.js:1487:1
Empty string passed to getElementById(). 3 LoginManager.shared.mjs:40:23
Character context menu initialized 
Object {  }
script.js:465:9
initializing Prompt Itemization Array on Startup script.js:473:9
Initializing Proxy Switcher SillyTavern-Proxy-Switcher.user.js:665:17
Proxy Switcher initialized SillyTavern-Proxy-Switcher.user.js:682:17
Unsupported language: en-us i18n.js:103:17
Deprecation warning: value provided is not in a recognized RFC2822 or ISO format. moment construction falls back to js Date(), which is not reliable across all browsers and versions. Non RFC2822/ISO date formats are discouraged. Please refer to http://momentjs.com/guides/#/warnings/js-date/ for more info.
Arguments: 
[0] _isAMomentObject: true, _isUTC: false, _useUTC: false, _l: undefined, _i: July 21, 2024 2:06am, _f: undefined, _strict: undefined, _locale: [object Object]
e/<@http://127.0.0.1:8000/lib/moment.min.js:1:3257
Nt@http://127.0.0.1:8000/lib/moment.min.js:1:21587
Pt@http://127.0.0.1:8000/lib/moment.min.js:1:22164
W@http://127.0.0.1:8000/lib/moment.min.js:1:22247
f@http://127.0.0.1:8000/lib/moment.min.js:1:207
parseTimestamp@http://127.0.0.1:8000/scripts/utils.js:695:15
timestampToMoment@http://127.0.0.1:8000/scripts/utils.js:671:34
addOneMessage@http://127.0.0.1:8000/script.js:2150:41
sendSystemMessage@http://127.0.0.1:8000/script.js:2565:18
firstLoadInit@http://127.0.0.1:8000/script.js:907:22
async*@http://127.0.0.1:8000/script.js:10829:11
e@http://127.0.0.1:8000/lib/jquery-3.5.1.min.js:2:30005
Deferred/then/l/</t<@http://127.0.0.1:8000/lib/jquery-3.5.1.min.js:2:30307
setTimeout handler*Deferred/then/l/<@http://127.0.0.1:8000/lib/jquery-3.5.1.min.js:2:30516
c@http://127.0.0.1:8000/lib/jquery-3.5.1.min.js:2:28294
add@http://127.0.0.1:8000/lib/jquery-3.5.1.min.js:2:28572
Deferred/then/<@http://127.0.0.1:8000/lib/jquery-3.5.1.min.js:2:30612
Deferred@http://127.0.0.1:8000/lib/jquery-3.5.1.min.js:2:31025
then@http://127.0.0.1:8000/lib/jquery-3.5.1.min.js:2:30541
S.fn.ready@http://127.0.0.1:8000/lib/jquery-3.5.1.min.js:2:31831
S.fn.init@http://127.0.0.1:8000/lib/jquery-3.5.1.min.js:2:25882
S@http://127.0.0.1:8000/lib/jquery-3.5.1.min.js:2:1024
@http://127.0.0.1:8000/script.js:8888:7
moment.js:287:20
Event emitted: settings_loaded_before eventemitter.js:101:17
Preset samplers order: 6,0,1,2,3,4,5 kai-settings.js:343:13
Preset samplers order: 1,5,0,2,3,4 nai-settings.js:504:13
Preset samplers order: 6,0,1,3,4,2,5 textgen-settings.js:531:13
Preset samplers order:  
Array(12) [ "temperature", "dynamic_temperature", "quadratic_sampling", "top_k", "top_p", "typical_p", "epsilon_cutoff", "eta_cutoff", "tfs", "top_a", … ]
textgen-settings.js:552:13
Preset samplers order:  
Array(6) [ "top_k", "tfs_z", "typical_p", "top_p", "min_p", "temperature" ]
textgen-settings.js:542:13
Claude model changed to claude-3-5-sonnet-20240620 openai.js:3860:17
Event emitted: chatcompletion_model_changed eventemitter.js:101:17
Event emitted: chatcompletion_source_changed eventemitter.js:101:17
poweruser.chat_display 0 power-user.js:1059:13
applying default chat power-user.js:1064:21
Window resize: 2560x1279 -> 2560x1279 power-user.js:3085:17
aborting MUI reset 0 power-user.js:3138:21
Window resize: 2560x1279 -> 2560x1279 power-user.js:3085:17
aborting MUI reset 0 power-user.js:3138:21
skipping movingUI state load power-user.js:1766:17
setting charlist state to... power-user.js:1734:17
..to list power-user.js:1738:26
Event emitted: settings_loaded_after eventemitter.js:101:17
value? openai script.js:6115:13
hiding settings? script.js:6117:17
returning full samplers array samplerSelect.js:310:17
Event emitted: worldinfo_settings_updated eventemitter.js:101:17
Event emitted: extensions_first_load eventemitter.js:101:17
skipping movingUI state load power-user.js:1766:17
Zoom: 2.00, X:2560, Y:1279, original: 5120x2558 power-user.js:3077:17
dragElement called for expression-holder RossAscends-mods.js:451:13
Summary set to:  Position: 0 Depth: 2 Role: 0 3 index.js:807:13
Using default TTS Provider settings system.js:116:21
SystemTTS: Settings loaded system.js:154:17
[USO] Button detection aborting due to timeout. injected.js:42:50363
Event emitted: undefined eventemitter.js:101:17
returning full samplers array samplerSelect.js:310:17
Event emitted: settings_loaded eventemitter.js:101:17
Failed to process stylesheet on initial load: DOMException: CSSStyleSheet.cssRules getter: Not allowed to access cross-origin stylesheet
    applyDynamicFocusStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:88
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:157
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:155
    firstLoadInit http://127.0.0.1:8000/script.js:910
    async* http://127.0.0.1:8000/script.js:10829
    jQuery 11
    <anonymous> http://127.0.0.1:8000/script.js:8888
dynamic-styles.js:159:21
Failed to process stylesheet on initial load: DOMException: CSSStyleSheet.cssRules getter: Not allowed to access cross-origin stylesheet
    applyDynamicFocusStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:88
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:157
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:155
    firstLoadInit http://127.0.0.1:8000/script.js:910
    async* http://127.0.0.1:8000/script.js:10829
    jQuery 11
    <anonymous> http://127.0.0.1:8000/script.js:8888
dynamic-styles.js:159:21
Failed to process stylesheet on initial load: DOMException: CSSStyleSheet.cssRules getter: Not allowed to access cross-origin stylesheet
    applyDynamicFocusStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:88
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:157
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:155
    firstLoadInit http://127.0.0.1:8000/script.js:910
    async* http://127.0.0.1:8000/script.js:10829
    jQuery 11
    <anonymous> http://127.0.0.1:8000/script.js:8888
dynamic-styles.js:159:21
Failed to process stylesheet on initial load: DOMException: CSSStyleSheet.cssRules getter: Not allowed to access cross-origin stylesheet
    applyDynamicFocusStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:88
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:157
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:155
    firstLoadInit http://127.0.0.1:8000/script.js:910
    async* http://127.0.0.1:8000/script.js:10829
    jQuery 11
    <anonymous> http://127.0.0.1:8000/script.js:8888
dynamic-styles.js:159:21
Failed to process stylesheet on initial load: DOMException: CSSStyleSheet.cssRules getter: Not allowed to access cross-origin stylesheet
    applyDynamicFocusStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:88
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:157
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:155
    firstLoadInit http://127.0.0.1:8000/script.js:910
    async* http://127.0.0.1:8000/script.js:10829
    jQuery 11
    <anonymous> http://127.0.0.1:8000/script.js:8888
dynamic-styles.js:159:21
Failed to process stylesheet on initial load: DOMException: CSSStyleSheet.cssRules getter: Not allowed to access cross-origin stylesheet
    applyDynamicFocusStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:88
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:157
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:155
    firstLoadInit http://127.0.0.1:8000/script.js:910
    async* http://127.0.0.1:8000/script.js:10829
    jQuery 11
    <anonymous> http://127.0.0.1:8000/script.js:8888
dynamic-styles.js:159:21
Failed to process stylesheet on initial load: DOMException: CSSStyleSheet.cssRules getter: Not allowed to access cross-origin stylesheet
    applyDynamicFocusStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:88
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:157
    initDynamicStyles http://127.0.0.1:8000/scripts/dynamic-styles.js:155
    firstLoadInit http://127.0.0.1:8000/script.js:910
    async* http://127.0.0.1:8000/script.js:10829
    jQuery 11
    <anonymous> http://127.0.0.1:8000/script.js:8888
dynamic-styles.js:159:21
[QR2] sets:  
Array [ {…} ]
index.js:17:51
[QR2] settings:  
Object { isEnabled: false, isCombined: false, isPopout: false, config: {…}, _chatConfig: undefined, onSave: undefined, onRequestEditSet: undefined }
index.js:17:51
[USO] Button detection aborting due to timeout. injected.js:42:50363
Event emitted: character_page_loaded eventemitter.js:101:17
Chat Completions: loading token cache tokenizers.js:115:17
Event emitted: settings_updated eventemitter.js:101:17
Registering preset manager for API: kobold preset-manager.js:93:21
Registering preset manager for API: novel preset-manager.js:93:21
Registering preset manager for API: openai preset-manager.js:93:21
Registering preset manager for API: textgenerationwebui preset-manager.js:93:21
Registering preset manager for API: context preset-manager.js:93:21
Registering preset manager for API: instruct preset-manager.js:93:21
User Agent 
Object { browser: {…}, os: {…}, platform: {…}, engine: {…} }
RossAscends-mods.js:729:13
setting pin class via checkbox state RossAscends-mods.js:804:17
setting pin class via checkbox state RossAscends-mods.js:814:17
LoadLocal -- userInput f-localStorage.js:8:13
Event emitted: app_ready eventemitter.js:101:17
[QR2] executing startup index.js:17:51
[QR2] /executing startup index.js:17:51
[QR2] executing queue (0 items) index.js:17:51
[QR2] /executing queue index.js:17:51
[QR2] READY index.js:17:51
SaveLocal -- NavOpened : true f-localStorage.js:5:13
Event emitted: settings_updated eventemitter.js:101:17
[USO] Button detection aborting due to timeout. injected.js:42:50363
saw no avatars script.js:1782:22
Event emitted: chat_id_changed eventemitter.js:101:17
Global world info has 0 entries world-info.js:3423:13
Character Sayuri's base world could not be found or is empty! Skipping... world-info.js:3380:17
Character Sayuri lore () has 0 world info entries world-info.js:3407:13
Sorted 0 world lore entries using strategy 1 world-info.js:3475:17
getting sprites list index.js:1278:13
[QR2] calling 
Object { functionToCall: async onChatChanged(chatIdx), args: (1) […] }
index.js:17:51
[QR2] CHAT_CHANGED 2024-7-21 @01h 18m 17s 652ms index.js:17:51

    setFloatingPrompt entered
    ------
    lastMessageNumber = 1
    metadata_keys.interval = 1
    metadata_keys.position = 1
    metadata_keys.depth = 4
    metadata_keys.role = 0
    ------
    authors-note.js:299:13
No default or locked persona set for this chat personas.js:865:17
Event emitted: chatLoaded eventemitter.js:101:17
Invalid URI. Load of media resource  failed. 127.0.0.1:8000
Empty string passed to getElementById(). 196 injected.js:42:17938
Event emitted: settings_updated eventemitter.js:101:17
Generate entered script.js:3195:13
Event emitted: generation_started eventemitter.js:101:17
Skipping extension interceptors for dry run script.js:3438:17
Core/all messages: 3/3 script.js:3456:13

    setFloatingPrompt entered
    ------
    lastMessageNumber = 1
    metadata_keys.interval = 1
    metadata_keys.position = 1
    metadata_keys.depth = 4
    metadata_keys.role = 0
    ------
    authors-note.js:299:13
Context size: 75904; WI budget: 18976 (max% = 25%, cap = 0) world-info.js:3525:13
Global world info has 0 entries world-info.js:3423:13
Character Sayuri's base world could not be found or is empty! Skipping... world-info.js:3380:17
Character Sayuri lore () has 0 world info entries world-info.js:3407:13
Sorted 0 world lore entries using strategy 1 world-info.js:3475:17
skipWIAN not active, adding WIAN script.js:3532:17
calling runGenerate script.js:3779:13
generating prompt script.js:3799:17
---calling setPromptString 0 script.js:3955:17
Event emitted: generate_after_combine_prompts eventemitter.js:101:17
Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user’s experience. For more help https://xhr.spec.whatwg.org/#sync-warning jquery-3.5.1.min.js:2:81767
Invalid message in collection undefined 2 openai.js:2764:25
Event emitted: chat_completion_prompt_ready eventemitter.js:101:17
Event emitted: online_status_changed eventemitter.js:126:17
Event emitted: settings_updated eventemitter.js:101:17
Switching to Charybdis - ant SillyTavern-Proxy-Switcher.user.js:345:17
Event emitted: online_status_changed eventemitter.js:126:17
value? openai script.js:6115:13
hiding settings? script.js:6117:17
returning full samplers array samplerSelect.js:310:17
Claude model changed to claude-3-5-sonnet-20240620 openai.js:3860:17
Event emitted: chatcompletion_model_changed eventemitter.js:101:17
Event emitted: online_status_changed eventemitter.js:126:17
Event emitted: chatcompletion_source_changed eventemitter.js:101:17
Claude model changed to claude-3-5-sonnet-20240620 openai.js:3860:17
Event emitted: chatcompletion_model_changed eventemitter.js:101:17
Event emitted: settings_updated eventemitter.js:101:17
Generate entered script.js:3195:13
Event emitted: generation_started eventemitter.js:101:17
Skipping extension interceptors for dry run script.js:3438:17
Core/all messages: 3/3 script.js:3456:13

    setFloatingPrompt entered
    ------
    lastMessageNumber = 1
    metadata_keys.interval = 1
    metadata_keys.position = 1
    metadata_keys.depth = 4
    metadata_keys.role = 0
    ------
    authors-note.js:299:13
Context size: 75904; WI budget: 18976 (max% = 25%, cap = 0) world-info.js:3525:13
Global world info has 0 entries world-info.js:3423:13
Character Sayuri's base world could not be found or is empty! Skipping... world-info.js:3380:17
Character Sayuri lore () has 0 world info entries world-info.js:3407:13
Sorted 0 world lore entries using strategy 1 world-info.js:3475:17
skipWIAN not active, adding WIAN script.js:3532:17
calling runGenerate script.js:3779:13
generating prompt script.js:3799:17
---calling setPromptString 0 script.js:3955:17
Event emitted: generate_after_combine_prompts eventemitter.js:101:17
Invalid message in collection undefined 2 openai.js:2764:25
Event emitted: chat_completion_prompt_ready eventemitter.js:101:17
Event emitted: message_swiped eventemitter.js:101:17
caught here 2 script.js:8143:45
Generate entered script.js:3195:13
Event emitted: generation_started eventemitter.js:101:17
Invalid URI. Load of media resource  failed. 127.0.0.1:8000
Running extension interceptors script.js:3429:17
Core/all messages: 2/3 script.js:3456:13

    setFloatingPrompt entered
    ------
    lastMessageNumber = 1
    metadata_keys.interval = 1
    metadata_keys.position = 1
    metadata_keys.depth = 4
    metadata_keys.role = 0
    ------
    authors-note.js:299:13
Context size: 75904; WI budget: 18976 (max% = 25%, cap = 0) world-info.js:3525:13
Global world info has 0 entries world-info.js:3423:13
Character Sayuri's base world could not be found or is empty! Skipping... world-info.js:3380:17
Character Sayuri lore () has 0 world info entries world-info.js:3407:13
Sorted 0 world lore entries using strategy 1 world-info.js:3475:17
skipWIAN not active, adding WIAN script.js:3532:17
calling runGenerate script.js:3779:13
generating prompt script.js:3799:17
---calling setPromptString 0 script.js:3955:17
Event emitted: generate_after_combine_prompts eventemitter.js:101:17
Event emitted: chat_completion_prompt_ready eventemitter.js:101:17
rungenerate calling API script.js:4164:17
pushed prompt bits to itemizedPrompts array. Length is now: 2 script.js:4208:17
Event emitted: chat_completion_settings_ready eventemitter.js:101:17
Event emitted: message_received eventemitter.js:101:17
Event emitted: character_message_rendered eventemitter.js:101:17
[QR2] calling 
Object { functionToCall: async onAiMessage(messageId), args: (1) […] }
index.js:17:51
[QR2] QR auto-execution suppressed for swiped message index.js:17:51
Stream stats: 268 tokens, 12.75 seconds, rate: 21.02 TPS script.js:3036:21
Event emitted: generation_ended eventemitter.js:101:17
Event emitted: message_received eventemitter.js:101:17
Event emitted: character_message_rendered eventemitter.js:101:17
[QR2] calling 
Object { functionToCall: async onAiMessage(messageId), args: (1) […] }
index.js:17:51
Chat Completions: saving token cache tokenizers.js:125:17
Auto-continue is disabled by user. script.js:4476:17
Generate entered script.js:3195:13
Event emitted: generation_started eventemitter.js:101:17
Skipping extension interceptors for dry run script.js:3438:17
Core/all messages: 3/3 script.js:3456:13

    setFloatingPrompt entered
    ------
    lastMessageNumber = 1
    metadata_keys.interval = 1
    metadata_keys.position = 1
    metadata_keys.depth = 4
    metadata_keys.role = 0
    ------
    authors-note.js:299:13
Context size: 75904; WI budget: 18976 (max% = 25%, cap = 0) world-info.js:3525:13
Global world info has 0 entries world-info.js:3423:13
Character Sayuri's base world could not be found or is empty! Skipping... world-info.js:3380:17
Character Sayuri lore () has 0 world info entries world-info.js:3407:13
Sorted 0 world lore entries using strategy 1 world-info.js:3475:17
skipWIAN not active, adding WIAN script.js:3532:17
calling runGenerate script.js:3779:13
generating prompt script.js:3799:17
---calling setPromptString 0 script.js:3955:17
Event emitted: generate_after_combine_prompts eventemitter.js:101:17
Invalid message in collection undefined 2 openai.js:2764:25
Event emitted: chat_completion_prompt_ready eventemitter.js:101:17
Generate entered script.js:3195:13
Event emitted: generation_started eventemitter.js:101:17
Skipping extension interceptors for dry run script.js:3438:17
Core/all messages: 3/3 script.js:3456:13

    setFloatingPrompt entered
    ------
    lastMessageNumber = 1
    metadata_keys.interval = 1
    metadata_keys.position = 1
    metadata_keys.depth = 4
    metadata_keys.role = 0
    ------
    authors-note.js:299:13
Context size: 75904; WI budget: 18976 (max% = 25%, cap = 0) world-info.js:3525:13
Global world info has 0 entries world-info.js:3423:13
Character Sayuri's base world could not be found or is empty! Skipping... world-info.js:3380:17
Character Sayuri lore () has 0 world info entries world-info.js:3407:13
Sorted 0 world lore entries using strategy 1 world-info.js:3475:17
skipWIAN not active, adding WIAN script.js:3532:17
calling runGenerate script.js:3779:13
generating prompt script.js:3799:17
---calling setPromptString 0 script.js:3955:17
Event emitted: generate_after_combine_prompts eventemitter.js:101:17
Invalid message in collection undefined 2 openai.js:2764:25
Event emitted: chat_completion_prompt_ready eventemitter.js:101:17
```

</details>

### Please tick the boxes

- [X] I have explained the issue clearly, and I included all relevant info
- [X] I have checked that this [issue hasn't already been raised](https://github.com/SillyTavern/SillyTavern/issues?q=is%3Aissue)
- [X] I have checked the [docs](https://docs.sillytavern.app/) ![important](https://img.shields.io/badge/Important!-F6094E)

**Comment by Cohee1207** [comment]
User Settings => Show {{char}}: in responses
User Settings => Show {{user}}: in responses

Make sure both are checked.

**Comment by Tomolasthill** [comment]
> User Settings => Show {{char}}: in responses User Settings => Show {{user}}: in responses

Damn, of all the testing I did. It never occurred to me that it was {{char}}/{{user}} cut-off. This helped. Thanks for the aid.

**Comment by Tomolasthill** [comment]
I closed this issue too early. I'd like to request to have "Show {{char}}/Show {{user}}: in responses" be reworked as a feature you **enable** to hide model Impersonations and **disabled** by **default**. As it currently stands, this being a default behavior is likely to cause more headaches then if it was on by default. I've had this issue some time now, but it was only after recently messing with CoT that it happened consistently enough to need further looking into.

I don't know what models the average SillyTavern user use, And I don't know how prominent model impersonations are on others/local models, but for popular proprietary models (OAI, Anthropic, Mistral etc.) I have not seen this issue since their earlier models. So from my perspective it seem better to leave this as a feature you enable to cut-off impersonations, rather then enable to not cut-off impersonations if that makes sense.

I also didn't see anything about "Show {{char}}/Show {{user}}: in responses" in the documentation, search gave nothing and going over the categories I don't see anything for user settings. So without asking or going over the code, it would be hard to know what the function does.

**Comment by Cohee1207** [comment]
There is never going to be a default configuration that satisfies everyone and will work for every single model and prompt. Defaults are usually optimized for the majority, which reportedly does not want to experience any form of model impersonation or explicit name prefixes.

Functions that are not explicitly stated in the docs (meaning the website) usually have on-hover tooltips. Reworking settings in a slightly more presentable and uncluttered state (à la VS Code settings menu) is out of the scope of this specific issue.

**Comment by Tomolasthill** [comment]
Understandable. I have no issue with this now that I understand what that function does. However, this function does not have a on-hover tooltips.  So without going in to test that function, It may not be intuitive to understand as is.

<s>Regardless, my issue has been solved. I won't inquire about this any further. I appreciate your help.</s>

Edit: So I've been told that by having "Show {{char}}/Show {{user}}: in response." un-ticked (disabled) then ST should only remove the prefix "{{char}}:" and nothing after it. My[ Streamable link](https://streamable.com/il99af) shows it completely cutting of the response from {{char}}: and after.

Is it suppose to cut off {{char}}/{{user}}: and everything after it as well?

**Comment by Wolfsblvt** [comment]
> However, this function does not have a on-hover tooltips.
Any suggestion on what hover tooltips would make sense there? Feedback welcome.

> Edit: So I've been told that by having "Show {{char}}/Show {{user}}: in response." un-ticked (disabled) then ST should only remove the prefix "{{char}}:" and nothing after it. My[ Streamable link](https://streamable.com/il99af) shows it completely cutting of the response from {{char}}: and after.

> 
> Is it suppose to cut off {{char}}/{{user}}: and everything after it as well?

So, I think this are two different things thrown together. And yes, it might be a bit confusing.
The checkbox itself only removes this from the display, meaning the rendered message.
There are separate settings for char and user prefixes being used inside the prompt itself. Text Comp has a checkbox, and pretty sure Chat Comp has something similar too.
And then lastly, there are the model stopping strings. And very often `\n{{user}}` or `\n{{user}}:` is part of the stopping strings. Which will make the generation stop completely, no matter what the other settings are.

Are you there is no interference with those? The regex and code for the display checkboxes look okay.

**Comment by Tomolasthill** [comment]
> Are you there is no interference with those? The regex and code for the display checkboxes look okay.
> The checkbox itself only removes this from the display, meaning the rendered message.

So, as my [Streamable link](https://streamable.com/il99af) shows. This is not a generation stop from a model stopping string. The rendered text is simply not shown after a certain point. Now, I decided to go back and test my Chain of Thought (CoT) prompt again. Here it is, now stripped down to showcase the issue:

<details>
 <summary>
 CoT
 </summary>
 
```
[Begin your response with an analysis of character(s) in scene. Use this template:

<Analysis>
Sayuri:
- Key traits: gender, personality, outfit
- Current portrayal vs. ideal portrayal
- Authenticity score: X/10 (explain briefly)

Vannian:
- Key character traits
- Authenticity score: X/10
</Analysis>

Be concise and direct. Reference character info from <name> tags when available. End response after the analysis.]
```

</details>

I've specifically added the {{char}}'s and {{user}}'s name there. This is what the `Show {{char}}/Show {{user}}: in response` Should remove. Just the `{{char}}:` and `{{user}}:` prefix, nothing after it. This I've managed to confirm.

First I've ran the CoT 10 times – with both boxes unchecked. Every single time the text was cut-off (not rendered) at mid-point. But the generation continues for a couple of second more as it was finishing the generation. So the CoT has a 100% failure rate. I noticed that it always begins the CoT correctly and without `{{char}}` prefix but fails when it reaches the `{{user}}:` portion. If I edit the finished generation, there is no `{{char}}:` prefix, only the bullet points for them.

So next I enabled `Show {{char}}: in responses`, this now showed the prefix + bullet points but still failed at the `{{user}}:` prefix point, I then disabled `Show {{char}}: in responses` and enabled `Show {{user}}: in responses`. And the generation finished while showing the `{{user}}:` prefix and the rest of the CoT but without `{{char}}:` prefix. Enabling {{char}} in response as well also showed the prefix correctly for {{char}} and the CoT worked as expected.

So it seems like the bug/issue lies with `Show {{user}}: in responses`. As not having that enabled removes `{{user}}:` prefix plus anything after from being rendered.

For reference. I am using a completely clean install of SillyTavern w/ a downloaded card and my own barebone preset with the CoT above, and no other modifications. Unlike my original post, this time I did update to the latest staging – (788a13d7) however, so there's that difference. So there are no custom settings, modifications or regexes applied here. Simply SillyTavern doing it's own thing.

Edit: I went through the effort of making a picture of 4 prompts. First 1 with both boxes unchecked, then trying them individually. The image doesn't show how it generates after the text rendering stops at mid-point, but you still see how `Show {{user}}: in responses` is bugged.
<details>
 <summary>
 The image - 350kb jpg.
 </summary>

![Show in responses](https://github.com/user-attachments/assets/8326dc9c-7f9e-4dc0-b363-8d9b6311944a)
</details>

**Comment by LenAnderson** [comment]
Well, looks like those two settings are a bit misleading and rather weirdly implemented.

- allow_name1_display (`{{user}}`)
  - Is applied during message cleanup (i.e., during streaming / after gen) and treated like a stopping string, if `{{user:}}` appears at the beginning of the message or directly after a newline, everything from that point on gets discarded. (Destructive) Since this is applied before anything is saved to the chat data, the removed text never reaches you at all.
  - The setting has no effect on impersonation prompts.
- allow_name2_display (`{{char}}`)
  - Is applied during message cleanup as (i.e., during streaming / after gen) and simply replaces any mention of `{{char}}:` (at the beginning of the message or directly after a newline) with nothing but doesn't throw away the rest of the message. (Destructive) Only the modified / stripped message text is saved to chat data.
  - Is applied during message formatting (i.e., when the chat data is turned into HTML for display). (Non-destructive) This means if you have chat messages that include this text, it will be stripped for display with the setting disabled, but will show up with the setting enabled. It is reversible, as it only affects how the message is displayed. The actual text remains in your saved chat data.
  - Is applied during message cleanup on impersonation prompts in the same way allow_name_1 is applied on non-impersonation prompts (i.e., stopping string, destructive).


It's another one of those places where the settings don't really reflect what's happening in the background. Who knows if there is (or has been at some point) a good reason for it...

**Comment by Wolfsblvt** [comment]
Thanks for looking deeper into this. I missed that specific part about `allow_name1_display`. Meh.
I *feel* like there is an actual reasoning behind that, but I can't say.

Likely something @Cohee1207 has to weigh in to tell us whether this is intended, and we should reword the toggles (or make tooltips), or whether this is an actual bug.

**Comment by Cohee1207** [comment]
I don't think toggles are wrongly worded or require fixing.

Not trimming the character name in the text and hiding it just visually has a negative effect: the model will start stacking names at the beginning of each reply. Same reason as they're starting to stack spaces if the space trimming is disabled (most models start the reply with a space).

**Comment by Wolfsblvt** [comment]
Are you sure?
When I look at the code Lenny mentioned, it looks very much like it doesn't just trim/remove the character name, but **everything** following this.
Here: https://github.com/SillyTavern/SillyTavern/blob/staging/public/script.js#L5281-L5286

**Comment by LenAnderson** [comment]
Indeed. The wording on the checkbox does not suggest that `{{user}}` will be treated like a stopping string.

**Comment by Wolfsblvt** [comment]
There are different options for doing that anyway.

**Comment by Cohee1207** [comment]
It needs to be trimmed and not just hidden visually otherwise it would get into prompts that follow. As already mentioned, it serves to prevent user impersonation in replies.

**Comment by Cohee1207** [comment]
(It's not simply adding a stop string, since not all backends can even support stop strings).

**Comment by Wolfsblvt** [comment]
Still, the setting itself implies it works the same as trimming Char:, which ONLY removes the char name from the display.

Maybe I just don't understand all implications of those settings.

**Comment by Tomolasthill** [comment]
> It needs to be trimmed and not just hidden visually otherwise it would get into prompts that follow. As already mentioned, it serves to prevent user impersonation in replies.
> (It's not simply adding a stop string, since not all backends can even support stop strings).

I understand, this is a measure to forcibly ensure that impersonation doesn't happen on models. So it's trimming away all content including the {{user}}: prefix from the response. So if `Show {{user}}: in responses` acted the same as `Show {{char}}: in responses`, The impersonation would still be inn but without the prefix, thus making the model more likely to continue writing/speaking/impersonate {{user}} in it's next response. This makes sense and confirms it's not a bug but an intended function.

However, as a user of SillyTavern and not a programmer. This means that the wording *should* be changed, docs updated or at least given a tool-tip to explain that this functions trims everything after the prefix. As it currently stands, it's simply confusing with different behavior. Especially since impersonation from popular proprietary models (GPT, Claude, Mistral etc..) can easily be stopped with a better prompt. While for local or weaker models, I can see this being a handy feature. But I assume local models can set stop-string anyways.

Once I learned what `Show {{user}}: in responses` did, my issue was easily resolved. But since there was no way for me to know outside asking here, I would have no idea how to fix this issue which has plagued me for a few weeks now. And let's be real, for most user looking at code is a no-go. And even if they do, this will look like gibberish to most.

![image](https://github.com/user-attachments/assets/b92a7340-2c44-44b4-a3cf-8a7e0e396321)

 So I still stand-by that this function should have a description/name or tool-tip to properly reflect what it does.
